cardano-node/test/Test/Cardano/Node/Chairman.hs

```diff
@@ -8,43 +9,69 @@ module Test.Cardano.Node.Chairman
+  startTime <- H.noteShow $ DTC.addUTCTime 60 currentTime -- 60 seconds into the future
 
-  H.createDirectoryIfMissing dbDir
-  H.createDirectoryIfMissing socketDir
+  -- Generate keys
+  void $ H.execCli
+    [ "genesis"
+    , "--genesis-output-dir", tempDir <> "/genesis"
+    , "--start-time", show @Int64 (floor (DTC.utcTimeToPOSIXSeconds startTime))
+    , "--protocol-parameters-file", base <> "/scripts/protocol-params.json"
+    , "--k", "2160"
+    , "--protocol-magic", "459045235"
+    , "--n-poor-addresses", "128"
+    , "--n-delegate-addresses", "7"
+    , "--total-balance", "8000000000000000"
+    , "--avvm-entry-count", "128"
+    , "--avvm-entry-balance", "10000000000000"
+    , "--delegate-share", "0.9"
+    , "--real-pbft"
+    , "--secret-seed", "2718281828"
+    ]
 
-  base <- H.getProjectBase
+  -- Launch cluster of three nodes
+  procResults <- forM [0..2] $ \i -> do
+    si <- H.noteShow $ show @Int i
+    dbDir <- H.noteShow $ tempDir <> "/db/node-" <> si
+    socketDir <- H.noteShow $ tempDir <> "/socket"
 
-  dirContents <- liftIO $ IO.listDirectory base
+    H.createDirectoryIfMissing dbDir
+    H.createDirectoryIfMissing socketDir
 
-  H.annotateShow $ dirContents
+    H.copyFile (baseConfig <> "/topology-node-" <> si <> ".json") (tempDir <> "/topology-node-" <> si <> ".json")
+    H.copyFile (baseConfig <> "/config-" <> si <> ".yaml") (tempDir <> "/config-" <> si <> ".yaml")
 
-  (Just hIn, _mOut, _mErr, hProcess) <- H.createProcess =<< H.procNode
-    [ "run"
-    , "--database-path", dbDir
-    , "--socket-path", socketDir <> "/node-2-socket"
-    , "--port", "3002"
-    , "--topology", base <> "/configuration/chairman/defaults/simpleview/topology-node-2.json"
-    , "--config", base <> "/configuration/chairman/defaults/simpleview/config-2.yaml"
-    , "--signing-key", base <> "/configuration/chairman/defaults/simpleview/genesis/delegate-keys.002.key"
-    , "--delegation-certificate", base <> "/configuration/chairman/defaults/simpleview/genesis/delegation-cert.002.json"
-    , "--shutdown-ipc", "0"
-    ]
+    (Just hIn, _mOut, _mErr, hProcess) <- H.createProcess =<< H.procNode
+      [ "run"
+      , "--database-path", dbDir
+      , "--socket-path", socketDir <> "/node-" <> si <> "-socket"
+      , "--port", "300" <> si <> ""
+      , "--topology", tempDir <> "/topology-node-" <> si <> ".json"
+      , "--config", tempDir <> "/config-" <> si <> ".yaml"
+      , "--signing-key", tempDir <> "/genesis/delegate-keys.00" <> si <> ".key"
+      , "--delegation-certificate", tempDir <> "/genesis/delegation-cert.00" <> si <> ".json"
+      , "--shutdown-ipc", "0"
+      ]
 
-  H.threadDelay 10000000
+    return (hIn, hProcess)
 
-  liftIO $ IO.hClose hIn
+  H.threadDelay 10000000
 
-  void $ H.waitForProcess hProcess
+  -- Signal for cluster to shutdown and wait for shutdown to complete
+  forM_ procResults $ \(hIn, _) -> liftIO $ IO.hClose hIn
+  forM_ procResults $ \(_, hProcess) -> void $ H.waitForProcess hProcess
```

Comment:
Can't that be done as:
```
forM_ procResults $ \(hIn, hProcess) -> do
    liftIO $ IO.hClose hIn
    void $ H.waitForProcess hProcess
```
and would that not make more sense?

Reply:
I think it's subtly different in the presence of exceptions, but in both cases, the result is not ideal.

Here, I'm assuming `hClose` doesn't throw, which would ensure I signal for all node process to exit before waiting for them to exit.

My assumption may be incorrect however.

Reply:
It looks like the worse that can happen if an exception is thrown during the test is any spawned processes will stick around until the test process exists.

I manually tested the situation where I kill the test process on MacOS and all the node processes disappear.

This might not be adequate if we launch multiple clusters and they try to open the same ports.

Or it might be if we only care about the first test failure, which I suspect may be the case.

Reply:
This PR removes the need to do this in the first place: https://github.com/input-output-hk/cardano-node/pull/1657

Perhaps that is a better approach?

Reply:
I think closing all the handles first might be a bit faster too because it there is latency in node shutdown code and closing all the handles means they can shutdown concurrently.  I haven't actually measured it though.

Reply:
Definitely the suggestion is more straightforward and can be preferred on that basis.

Reply:
The other PR https://github.com/input-output-hk/cardano-node/pull/1657 uses `cleanupProcess`, which might be less polite, but I suspect non-politeness is actually fine because every test is isolated with its own database directory.  But we maybe want at least one polite test.